examples/huggingface/main.py

```diff
@@ -0,0 +1,320 @@
+        action="store_true",
+        required=False,
+    )
+    parser.add_argument(
+        "-lr", "--learning_rate", help="Learning rate", type=float, required=False
+    )
+    parser.add_argument(
+        "-e",
+        "--num_train_epochs",
+        help="Number of training epochs",
+        type=int,
+        required=False,
+    )
+    parser.add_argument(
+        "-b",
+        "--per_device_train_batch_size",
+        help="Per device batch size",
+        type=int,
+        required=False,
+    )
+    parser.add_argument(
+        "-opt",
+        "--optim",
+        help="Optimizer (one of: adamw_hf, adamw_torch, adamw_apex_fused, or adafactor",
+        type=str,
+        required=False,
+    )
+    parser.add_argument(
+        "-wd",
+        "--weight_decay",
+        help="Weight decay for AdamW optimizer",
+        type=float,
+        required=False,
+    )
+    parser.add_argument(
+        "-b1",
+        "--adam_beta1",
+        help="beta1 hyperparameter for AdamW optimizer",
+        type=float,
+        required=False,
+    )
+    parser.add_argument(
+        "-b2",
+        "--adam_beta2",
+        help="beta2 hyperparameter for AdamW optimizer",
+        type=float,
+        required=False,
+    )
+    parser.add_argument(
+        "-eps",
+        "--adam_epsilon",
+        help="epsilon hyperparameter for AdamW optimizer",
+        type=float,
+        required=False,
+    )
+    parser.add_argument(
+        "-log", "--logfile", help="Log file name and path", type=str, required=False
+    )
+    args = parser.parse_args()
+    return vars(args)
+
+
+def set_training_args(training_args, args):
+    for argname, argvalue in args.items():
+        if argvalue is not None:
+            setattr(training_args, argname, argvalue)
+    return training_args
+
+
+class GPUMemoryCallback(TrainerCallback):
+    def on_epoch_end(self, args, state, control, **kwargs):
+        print(
+            "GPU mem: Tot - ",
+            torch.cuda.get_device_properties(0).total_memory,
+            "res - ",
+            torch.cuda.memory_reserved(0),
+            "used - ",
+            torch.cuda.memory_allocated(0),
+        )
+
+
+def get_free_gpu():
+    for i in range(torch.cuda.device_count()):
+        gpu_procs_str = torch.cuda.list_gpu_processes(i)
+        if "no processes are running" in gpu_procs_str:
+            return i
+    return None
+
+
+@hydra.main(config_path=".", config_name="config")
+def main(cfg: DictConfig) -> float:
+    print("args", cfg)
+
+    # Get command line arguments and apply hyperparaters to training arguments
+    args = cfg.args
+
+    # Logger setup
+    logfile = args["logfile"] or "translation_hf.log"
+
+    logging.basicConfig(filename=logfile, level=logging.INFO)
+    logger = logging.getLogger()
+
+    # Get a GPU if available
+    if torch.cuda.is_available():
+        device = f"cuda:{get_free_gpu()}"
+    else:
+        device = "cpu"
+
+    logger.info("Compute device: %s", device)
```

Comment:
`device` is only used for printing. Are we sure the trainer is using the proper device?

Reply:
This was part of Sean's code. Seems like he only wanted to print it, because even if we wanted to set it for Huggingface, looks like it's not working yet on Huggingface's side : https://discuss.huggingface.co/t/setting-specific-device-for-trainer/784/6

Reply:
Maybe setting a specific GPU does not work, but at least it can use GPUs. Maybe just add a comment above this logger.info stating that we only use `device` to print what device we assume HuggingFace should be using by default.

Reply:
[303e6a0](https://github.com/Epistimio/orion/pull/1040/commits/303e6a0a4ccd1092ddf83a738d2368d5244093d4)